### PR TITLE
fix: force `GoldClusterizer` to cluster every sample in 1 unique

### LINKED ISSUE #226: Force `GoldClusterizer` to cluster every sample in 1 unique single cluster
**Context & Problem**

In the output of a `GoldClusterizer` (from `cluster_in_dataset` or `cluster_in_table`), every vectors are assigned to 1 cluster. Depending on the situation, multiple vectors might come from the same sample and asisgned to different clusters. 

For some situation, we might want to assign all the vectors of the same sample in the same cluster.

**Proposed Solution**

In `GoldClusterizer`, we add an new boolean attribute `force_same_cluster`. When it is True, we update the cluster of all the row of the same sample (using `idx`) instead of just the unique. `idx_vector` row. To do that we need to extract sample idx associated to the vector idx.

### PR DESCRIPTION
**Pull request by Mr-Neutr0n**
Fixes #226

Added a `force_same_cluster` boolean parameter to `GoldClusterizer` (default `False`); when enabled, `_cluster_label` maps vector indices to sample `idx` values and updates all rows for each sample so every vector from the same sample shares one cluster.

Local tests pass.

---
This change was prepared with AI assistance under human direction and review.

---
## Summary by cubic
Fixes #226 by adding an opt-in flag so `GoldClusterizer` assigns one cluster per sample instead of per vector. When enabled, all vectors of a sample (`idx`) share the label of the last processed vector of that sample.

- New parameter `force_same_cluster=False`; when True, `_cluster_label` keys off `idx` instead of `idx_vector`.
- Side effect: a sample’s final label depends on processing order; later vectors can overwrite earlier labels.
- Migration: the `idx` column must exist in both `cluster_from` and `cluster_table` before enabling.
- Default behavior unchanged; enabling removes within-sample label variance.

<sup>Written for commit 7e09bb867bd21893c70aa2f5c3ff07fdefecf344. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/goldener-data/goldener/pull/306?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: goldener/clusterize.py [hunk 1]
```diff
@@ -949,18 +955,22 @@ def _cluster_label(
...
 
+        idx_key = "idx" if self.force_same_cluster else "idx_vector"
+        idx_col = get_expr_from_column_name(cluster_from, idx_key)
+        idx_expr = get_expr_from_column_name(cluster_table, idx_key)
+
         for chunk_indices in chunk_assignment:
             to_cluster_from = cluster_from.where(
                 cluster_from.idx_vector.isin(chunk_indices)
-            ).select(vectorized_col, cluster_from.idx_vector)
+            ).select(vectorized_col, idx_col)
```
Comment (Yann-CV) [inline goldener/clusterize.py]: chunk indices are computed from to_cluster_vector_indices coming from `idx_vector`.

Even though, this is not the best solution (IMO would be to aggregate all the vectors of same sample and do the clustering from it) for now, we just want to keep simple and only change line 996 here.

File: goldener/clusterize.py [hunk 2]
```diff
@@ -949,18 +955,22 @@ def _cluster_label(
             random_state=self.random_state,
         )
 
+        idx_key = "idx" if self.force_same_cluster else "idx_vector"
+        idx_col = get_expr_from_column_name(cluster_from, idx_key)
+        idx_expr = get_expr_from_column_name(cluster_table, idx_key)
+
```
Comment (Yann-CV) [inline goldener/clusterize.py]: can be removed

File: goldener/clusterize.py [hunk 3]
```diff
@@ -949,18 +955,22 @@ def _cluster_label(
...
+            ).select(vectorized_col, idx_col)
 
             if to_cluster_from.count() == 0:
                 raise ValueError("Unexpected empty chunk for clustering.")
 
             to_cluster_for_chunk = [
                 (
                     torch.from_numpy(sample[self.vectorized_key]),
-                    torch.tensor(sample["idx_vector"]).unsqueeze(0),
+                    torch.tensor(sample[idx_key]).unsqueeze(0),
```
Comment (Yann-CV) [inline goldener/clusterize.py]: can eb removed

File: goldener/clusterize.py [hunk 4]
```diff
@@ -983,7 +993,7 @@ def _cluster_label(
                 set_value_to_idx_rows(
                     table=cluster_table,
                     col_expr=cluster_col,
-                    idx_expr=cluster_table.idx_vector,
+                    idx_expr=idx_expr,
```
Comment (Yann-CV) [inline goldener/clusterize.py]:
```suggestion
                    idx_expr="idx" if self.force_same_cluster else "idx_vector",
```

File: goldener/clusterize.py [hunk 5]
```diff
@@ -253,6 +253,8 @@ class GoldClusterizer:
             when using `cluster_in_dataset`. Default is False.
         max_batches: Optional maximum number of batches to process. Useful for testing on a small subset of the dataset.
         random_state: Optional random state for reproducibility during chunk assignment.
+        force_same_cluster: If True, all vectors from the same sample (same `idx`) are assigned
+            to the same cluster. Default is False.
```
Comment (Yann-CV) [inline goldener/clusterize.py]:
```suggestion
        force_same_cluster: If True, all vectors from the same sample (same `idx`) are assigned
            to the same cluster. The assignement to a cluster depends on the last clusterized vector of the vector (suboptimal but easiest solution). Default is False.
```

**Review by Yann-CV — CHANGES_REQUESTED** [review]
Thanks for your contribution!

I have made some proposal as comments, let's keep it simple for now! In addition to the suggestion, I guees it worths to add a unit test for that

**Comment by Mr-Neutr0n** [comment]
Applied your suggestions (ddf82e7): dropped the idx_key/idx_col/idx_expr plumbing and the select/tensor changes, and line 996 now switches the update target inline (`cluster_table.idx` when `force_same_cluster` else `cluster_table.idx_vector`), keeping the chunking exactly as before. Updated both docstrings with your wording. Added `test_cluster_in_table_force_same_cluster` (10 samples x 3 vectors, chunked, asserts every sample lands in exactly one cluster) plus a companion test pinning the default per-vector behavior. All 38 tests in test_clusterize.py pass.

File: goldener/clusterize.py [hunk 6]
```diff
@@ -983,7 +991,11 @@ def _cluster_label(
                 set_value_to_idx_rows(
                     table=cluster_table,
                     col_expr=cluster_col,
-                    idx_expr=cluster_table.idx_vector,
+                    idx_expr=(
```
Comment (Yann-CV) [inline goldener/clusterize.py]: apologies I guess I missed the fact that you should extract as well the right idx column in 
```
idx_expr = (
    cluster_table.idx 
    if self.force_same_cluster
    else cluster_table.idx_vector
)
to_cluster_from = cluster_from.where(
    cluster_from.idx_vector.isin(chunk_indices)
).select(vectorized_col, idx_expr)

...

set_value_to_idx_rows(
    table=cluster_table,
    col_expr=cluster_col,
    idx_expr=idx_expr,
    indices=set(indices_in_cluster),
    value=cluster_idx,
)


```

Comment (Yann-CV) [reply goldener/clusterize.py]: to_cluster_for_chunk ise either idx or idx_vector

File: tests/test_clusterize.py [hunk 7]
```diff
@@ -533,6 +533,79 @@ def test_cluster_in_table_with_chunk_larger_than_total(self):
...
+            table_path=cluster_path,
+            clustering_tool=GoldRandomClusteringTool(random_state=0),
+            allow_existing=True,
+            chunk=10,
+            force_same_cluster=True,
+        )
+
+        cluster_table = clusterizer.cluster_in_table(src_table, n_clusters=4)
+
+        assert cluster_table.count() == 30
```
Comment (Yann-CV) [inline tests/test_clusterize.py:580]: you can use `get_cluster_indices` and check that no samples are shared between the set

File: tests/test_clusterize.py [hunk 8]
```diff
@@ -533,6 +533,79 @@ def test_cluster_in_table_with_chunk_larger_than_total(self):
...
+        sample_clusters: dict[int, set] = {}
+        for row in cluster_table.select(
+            cluster_table.idx, cluster_table[clusterizer.cluster_key]
+        ).collect():
+            assert row[clusterizer.cluster_key] is not None
+            sample_clusters.setdefault(row["idx"], set()).add(
+                row[clusterizer.cluster_key]
+            )
+
+        assert set(sample_clusters) == set(range(10))
```
Comment (Yann-CV) [inline tests/test_clusterize.py]: I would not expect this

File: tests/test_clusterize.py [hunk 9]
```diff
@@ -533,6 +533,79 @@ def test_cluster_in_table_with_chunk_larger_than_total(self):
...
+        assert all(len(clusters) == 1 for clusters in sample_clusters.values())
+
+    def test_cluster_in_table_without_force_keeps_per_vector_labels(self):
```
Comment (Yann-CV) [inline tests/test_clusterize.py]: not sure I get this test

**Review by Yann-CV — CHANGES_REQUESTED** [review]
Sorry I missed something in the first review! let me know if you need help on this.

Thanks for the unit test but I guess you need to make sure they match the intended purposes

**Comment by Mr-Neutr0n** [comment]
Addressed the review: `force_same_cluster` now selects the source `idx`, carries it through clustering, and updates the cluster table using the same key. I also replaced the broad test with `get_cluster_indices` checks that prove sample sets are disjoint across clusters, and removed the unrelated per-vector test. All 37 clusterizer tests pass, plus Ruff and mypy. https://github.com/Mr-Neutr0n/goldener/commit/75cc43a

**Review by Yann-CV — APPROVED** [review]
LGTM great work!

Hope to review you again soon!

**Comment by Yann-CV** [comment]
@Mr-Neutr0n just need to fix the conflicts before merging!

By the way, I am meeting with the different contributing to give them more info about Goldener and long term vision. Would you be interested?

**Comment by Mr-Neutr0n** [comment]
@Yann-CV Yes please. harikp2002@gmail.com is me.

will fix the conflicts.

**Comment by Yann-CV** [comment]
@Mr-Neutr0n what is your timezone? Have you any preference on time and date?

**Comment by Mr-Neutr0n** [comment]
> what is your timezone? Have you any preference on time and date?
@Yann-CV  8:30PM PST to 11:30AM PST. I am currently in GMT + 5:30

**Comment by Mr-Neutr0n** [comment]
@Yann-CV The conflict with current main is resolved in `7e09bb8`. I also restored the default per-vector multi-vector regression test requested by Cubic. Full test suite: 443 passed. Mypy, Ruff, formatting, pre-commit, and diff checks pass. GitHub now reports the PR approved, mergeable, and all checks green. Please merge it when convenient.

**Comment by Yann-CV** [comment]
@Mr-Neutr0n I am sending you an invite, let me know if it does not suit you ;-)
